### PR TITLE
[Inductor] Add triton.autotune support for user defined triton kernels with complex grids

### PR DESCRIPTION
**Pull request by oulgen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112292
* __->__ #112290

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_higher_order_ops/triton_kernel_wrap.py [hunk 1]
```diff
@@ -79,7 +79,23 @@ def __init__(self):
...
+        def grid_fn(meta):
+            configs = kernel.configs
+            assert len(grid) == len(configs)
+            for i, (grid_val, config) in enumerate(zip(grid, configs)):
+                guards = [
+                    f"meta['{name}'] == {val}" for name, val in config.kwargs.items()
+                ]
+                guards = " and ".join(guards)
+                if eval(guards):
+                    return grid[i]
```
Comment (jansel) [inline torch/_higher_order_ops/triton_kernel_wrap.py:96]: Configs aren't actually guaranteed to be unique, since they can differ in num_stages/num_warps (but not kwargs).  This might be fine though, since I think the grids will be the same for duplicates.

Calling `eval()` is very expensive in python, and this function is O(num-configs).

This this could be implemented more efficiently using a dict.

File: torch/_inductor/codegen/wrapper.py [hunk 2]
```diff
@@ -496,10 +496,28 @@ def generate_extern_kernel_out(self, output_view, codegen_reference, args, kerne
...
+            grid_wrapper = IndentedBuffer()
+            grid_wrapper.writeline(f"def grid_wrapper_for_{kernel_name}(meta):")
+            assert len(grid) == len(configs)
+            with grid_wrapper.indent():
+                for i, (grid_val, conf) in enumerate(zip(grid, configs)):
+                    guards = [
+                        f"meta['{name}'] == {val}" for name, val in conf.kwargs.items()
+                    ]
+                    guards = " and ".join(guards)
+                    grid_wrapper.writeline(f"if {guards}: return grid({grid[i]})(meta)")
```
Comment (jansel) [inline torch/_inductor/codegen/wrapper.py:513]: Could we share code with the implementation above?  I think we can preprocess this into a dict, so the body is more like:
```
def grid(meta):
  return precomputed_grids[(meta["XBLOCK], meta["YBLOCK])]
```

File: torch/_inductor/triton_heuristics.py [hunk 3]
```diff
@@ -1191,9 +1191,10 @@ def foreach(meta, num_warps, filename=None):
 
 def grid(*numels):
     """Helper function to compute triton grids"""
-
     if len(numels) == 1 and isinstance(numels[0], tuple):
-        numels = numels[0]
+        # when passed as a single tuple, calling convention order is not
+        # followed, so we need to reverse to match calling conversion order
+        numels = numels[0][::-1]
```
Comment (jansel) [inline torch/_inductor/triton_heuristics.py:1197]: I don't understand why this change is needed, aren't we not using this function for triton kernels?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/triton_heuristics.py (hunk 3 above)
Comment (oulgen) [reply torch/_inductor/triton_heuristics.py:1197]: cached_autotune uses the launder function which requires everything to be 3tuple since it unpacks.

https://github.com/pytorch/pytorch/blob/c14c4efc0eb97ccc471ac6cbdcc659417227ae97/torch/_inductor/triton_heuristics.py#L316-L319

I was using this function to convert 1tuple or 2tuple into 3 tuple. I can either write my own function to do this or is there a more pythonic way to do this converstion?

File: torch/_higher_order_ops/triton_kernel_wrap.py (hunk 1 above)
Comment (oulgen) [reply torch/_higher_order_ops/triton_kernel_wrap.py:96]: > This might be fine though, since I think the grids will be the same for duplicates

grid computation doesn't use num_warps and num_stages so duplicates are fine, just more branches.

File: torch/_inductor/codegen/wrapper.py (hunk 2 above)
Comment (oulgen) [reply torch/_inductor/codegen/wrapper.py:513]: One edge case I see here is
```
@triton.autotune(
        configs=[
            triton.Config(
                {"BLOCK_SIZE_X": 128, "BLOCK_SIZE_Y": 128}, num_stages=3, num_warps=8
            ),
            triton.Config(
                {"BLOCK_SIZE_X": 64}, num_stages=3, num_warps=8
            ),
        ],
        key=[],
    )
```
where configs are not of equal size. Is this something I should disallow?

Comment (jansel) [reply torch/_inductor/codegen/wrapper.py:513]: I think we can just disallow this.

File: torch/_inductor/triton_heuristics.py (hunk 3 above)
Comment (jansel) [reply torch/_inductor/triton_heuristics.py:1197]: Since we are generating the grid function, can't we just make it return a 3-tuple?

**Comment by oulgen** [comment]
Explain offline to @jansel that using a dict imposes unnecessary restrictions. updated the code to only have a single exec in eager mode, and share code between eager and inductor.

**Review by jansel — APPROVED** [review]

File: torch/_dynamo/variables/functions.py [hunk 4]
```diff
@@ -689,31 +689,59 @@ def __init__(self, kernel, kernel_idx, grid, **kwargs):
...
+            # inductor expects all grids to be 3-tuple so lets make it
+            if len(grids[i]) == 1:
+                grids[i] = (grids[i][0], 1, 1)
+            elif len(grids[i]) == 2:
+                grids[i] = (grids[i][0], grids[i][1], 1)
+            elif len(grids[i]) > 3:
+                raise Unsupported("Grid can have at most rank 3")
+
+        assert len(grids) != 0
+        if len(set(grids)) == 1:
```
Comment (ezyang) [inline torch/_dynamo/variables/functions.py:742]: This test causes SymInt specialization when there is SymInt in the grids, as SymInts are not hashable ordinarily.

Comment (oulgen) [reply torch/_dynamo/variables/functions.py:742]: Is there a better way to do this or should we kill this optimization?

Comment (oulgen) [reply torch/_dynamo/variables/functions.py:742]: Lets just kill it since this optimization is probably overkill anyway
